### ISSUE TITLE
Function links in stack backtraces

### ISSUE THREAD

**Issue by alygin** [intellij-rust/intellij-rust#821]
For Java, Kotlin and perhaps other JVM languages, IDEA provides a nice stacktrace presentation. It adds links to exact positions in the source code for every line, it separates our methods calls from calls to external libraries, it supports folding etc.

I've prototyped a similar feature for Rust. So far, it makes stack backtraces look like this:

<img width="641" alt="screen shot 2016-12-08 at 00 26 22" src="https://cloud.githubusercontent.com/assets/2101250/20987268/0587a2e4-bcdd-11e6-9ca6-5731dae60e69.png">

Here, the usual hyperlinks lead to the source code of the project we are working on. The grayed hyperlinks lead to external crates. And those not linked are from a special skip list. Every link leads to the declaration of the corresponding function. Trailing hashcodes are dimmed a bit to reduce noise.

But before moving further with this feature, I'd like to discuss it here. What do you think, does it worth implementing at all? Any ideas on how to improve it?

There's also a problem I wasn't able to solve quickly &mdash; it seems that `RustResolveEngine.resolve(path, pivot)` doesn't work for struct methods, it only works for pure and associated functions. Am I right?

**Comment by alygin** [comment]
And to be objective, here are my arguments against this feature:

1. If the described problem with resolving methods is really unsolvable now, we'll have many unlinked lines which will be rather confusing.
2. The mechanism is not sophisticated enough to provide links to the exact call positions, it only leads to method declarations. The developer still needs to search for the exact calls.
3. On Linux, stack backtraces already come with the exact source code links (we now parse them and wrap into hyperlinks). Maybe we just need to wait until such feature comes to other platforms.

The feature might possibly be a bit different &mdash; just highlight parts of stack backtraces to make its analysis easier, but don't add hyperlinks. Or to add hyperlinks only on non-linux platforms.

**Comment by matklad** [comment]
Sorry, looks like I've missed notification about this issue :( 

We surely do need to hyperlink qualified names in stacktraces and in general we should have an API like `resolveQualifiedNameString(path: String) -> RustPathNamedElement`. 


> it only works for pure and associated functions.

Yes, and this is a bug I think.

**Comment by alygin** [comment]
OK, then I'll try to tackle the bug and move the function that searches a `RustPathNamedElement` into a utility class (into `RustResolveUtil`?).

**Comment by matklad** [comment]
Into `RustResolveEngine`. `RustResovleUtil` should ideally be dead, but I can;t

**Comment by matklad** [comment]
As for the bug,I think you need to consider all methods in the resolve, and not only the static ones:

https://github.com/intellij-rust/intellij-rust/blob/bb9acde13645abef5c8da9422719794bb1118da4/src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt#L389-L392

That is, the following is a valid Rust code:

```
struct S;

impl S { fn test(&self) {} }

fn main() {
    S::test(&S);
}
```

**Comment by alygin** [comment]
Sure, that was exactly my point, we must resolve everything for this feature to be useful and not confusing.

**Comment by alygin** [comment]
Oh, you meant the exact place in the code. I misunderstood your comment in a hurry.

**Comment by alygin** [comment]
Closed via #827
